### DISCUSSION TITLE
Integrate GA(Genetic Algorithms) inside CP-SAT as a subsolver?

### DISCUSSION THREAD

**Discussion by C8504** [google/or-tools#2805] [CP-SAT questions] [answered]
Hi Experts,

Do we have any plan to integrate GA inside CP-SAT as a subsolver?
If no, could you please share any ideas or guidelines to integrate?

Thanks

**Comment by lperron** [comment]
no.

**Reply by lperron** [reply]
and no special idea/interest in doing so. I do not believe it is a good idea.

**Reply by C8504** [reply]
Thanks, while could we do something like recombination between 2 feasible solutions during subsolver like LNS is searching.
For example, we have 2 solutions by LNS,
001010001
011000110
then we fixed the same 0 and 1 above to create
`0*10*0***`
then we use above fix and optimize to create a new neighborhood?

Thanks again!

**Comment by lperron** [comment, marked as answer]
I know exactly what a GA is :-)

It is just that the probability of a recombined solution is very low (= ~0)
with a direct encoding. So it is not worth it. Trust me, I spend a lot of
time working on this potential improvements to the search.

GA works well when using an indirect encoding. But this requires problem
specific information to be efficient, so it should be done on top of the
solver.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by C8504** [reply]
Thanks Laurent！ ^_^

**Comment by gregy4** [comment]
I can’t talk about cp-sat but for constraint solver it was helpful (variation of search heuristic) but too pricey to get good solution fast.

Integration of another solver with cp-sat is still not possible. Last time Laurent told me that communication between search workers is not ready to add this functionality (should be stabilized). For me, it is for sure a way to go.
